crates/autopilot/src/run_loop.rs

```diff
@@ -450,45 +408,77 @@ impl RunLoop {
 
-        let order_uids = auction.orders.iter().map(|o| OrderUid(o.uid.0)).collect();
+        let order_uids = auction.orders.iter().map(|o| OrderUid(o.uid.0));
         self.persistence
             .store_order_events(order_uids, OrderEventLabel::Ready);
 
-        let start = Instant::now();
-        futures::future::join_all(self.drivers.iter().map(|driver| async move {
-            let result = self.solve(driver, request).await;
-            let solutions = match result {
-                Ok(solutions) => {
-                    Metrics::solve_ok(driver, start.elapsed());
-                    solutions
-                }
-                Err(err) => {
-                    Metrics::solve_err(driver, start.elapsed(), &err);
-                    if matches!(err, SolveError::NoSolutions) {
-                        tracing::debug!(driver = %driver.name, "solver found no solution");
-                    } else {
-                        tracing::warn!(?err, driver = %driver.name, "solve error");
-                    }
-                    vec![]
-                }
-            };
-
-            solutions.into_iter().filter_map(|solution| match solution {
-                Ok(solution) => {
-                    Metrics::solution_ok(driver);
-                    Some(Participant { driver, solution })
-                }
-                Err(err) => {
-                    Metrics::solution_err(driver, &err);
-                    tracing::debug!(?err, driver = %driver.name, "invalid proposed solution");
-                    None
-                }
-            })
-        }))
+        let mut solutions = futures::future::join_all(
+            self.drivers
+                .iter()
+                .map(|driver| self.solve(driver, request)),
+        )
         .await
         .into_iter()
         .flatten()
-        .collect()
+        .collect::<Vec<_>>();
+
+        // Shuffle so that sorting randomly splits ties.
+        solutions.shuffle(&mut rand::thread_rng());
+        solutions.sort_unstable_by_key(|participant| participant.solution.score().get().0);
+
+        // Make sure the winning solution is fair.
+        while !Self::is_solution_fair(solutions.last(), &solutions, auction) {
+            let unfair_solution = solutions.pop().expect("must exist");
+            warn!(
+                invalidated = unfair_solution.driver.name,
+                "fairness check invalidated of solution"
+            );
+        }
+        self.report_on_solutions(&solutions, auction);
+
+        solutions
+    }
+
+    /// Records metrics, order events and logs for the given solutions.
+    /// Expects the winning solution to be the last in the list.
+    fn report_on_solutions(&self, solutions: &[Participant<'_>], auction: &domain::Auction) {
+        let Some(winner) = solutions.last() else {
+            // no solutions means nothing to report
+            return;
+        };
+
+        solutions.iter().for_each(|solution| {
```

Comment:
This would create a log per solution, wouldn't it better to have all the solutions in the same log? that way we have them all together and not scattered with the other logs, easier to find.

Reply:
I think there are trade-offs to both. For example searching something like `(log: "oneinch" or log: "zeroex") and log: "<UID>"` would be way nicer with individual logs IMO.
Given that the grouped log is more complex to implement and it's not a clear usability improvement IMO I would prefer to leave it as is until it turns out to be annoying in practice.